include/Gaffer/FileSystemPath.h

```diff
@@ -82,6 +82,14 @@ class GAFFER_API FileSystemPath : public Path
 		// a FileSequence.
 		IECore::FileSequencePtr fileSequence() const;
 
+		/// Sets the path root and names in generic format from an OS native path string
+		void setFromString(const std::string &string) override;
+
+		// Returns the path in generic format
+		std::string string() const override;
+		// Returns the path converted to the OS native format
+		std::string nativeString() const;
```

Comment:
I see your point about symmetry. This may be a bit of a boat-rocker, but what about deprecating `setFromString()` and adding a new `set( const std::string &string)` function to take its place?

> When a user types a Windows-formatted string into a widget, and we want to turn it into a generic-formatted string without the widget knowing anything about FileSystemPath?

Yes, it looks like the system you describe would require a little more care when using `setFromString()` to use `setFromNativeString()` if it's a file path. Doing a global search for `setFromString` I don't see all that many cases where it would need to be changed though.

Reply:
OK, in that case I don't think it's  reasonable to expect all the Path clients to do an `if isinstance( p, FilesSystemPath ) : setFromNativeString()` switcheroo. The whole point of having a Path abstraction is to avoid that sort of thing. Things are feeling much beter now that `string()` isn't virtual at least, so I guess we'll have to go with `setFromString()` being virtual in some way (and removing `setFromNativeString()`).

One I thing I just realised though is that the current implementation of `setFromNativeString()` can emit `pathChanged` twice (once for the root and once for the names), with the first emission exposing an incomplete intermediate state to callers. We'd have the same problem if we simply made `setFromString()` virtual. Perhaps we need this?

```
public :

    // Non-virtual as before, but now implemented using `decomposeString()`.
    void setFromString( const std::string &s );

private :

    // Can be overridden by FileSystemPath to do what we need. But signalling will still
    // be done in `setFromString()`.
    virtual decomposeString( const std::string &s, InternedString &root, Names &names ) const;
```

Either way it's an ABI break, but I think this does a better job of the signalling and gives us a bit more flexibility...

Reply:
I'm not sure I got the update in 1a06f2b4ccd5fcce4594a3ccf2fccc83b2066c7d right based on what we talked about this morning. At the time I think you were expecting to be able to leave `m_root` and `m_names` `private`, whereas I've moved them to `protected` here.

Should I be returning the root and names from `setRootAndNames()` and setting the actual `m_root` and `m_names` in `setFromString()`?

Reply:
> Should I be returning the root and names from setRootAndNames() and setting the actual m_root and m_names in setFromString()?

I think so. I was envisaging `rootAndNames()` just being a better-named version of the `decomposeString()` function I originally sketched out above :

```virtual void rootAndNames( const std::string &s, InternedString &root, Names &names ) const;```

The benefit of this is that `setFromString()` is still able to emit `pathChangedSignal()` only when this actually results in a name. And `rootAndNames()` can be potentially used in places where we don't want to modify the path at all.